Introduction to Performance Optimization

IanTeo opened the pull request.

IanTeo committed:
Add introduction to performance optimization

IanTeo committed:
Add descriptions to links and improve formatting

ablyx commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+## Overview
+
+Computer performance can be defined as the rate of work accomplished by a computer system. Even if execution time is not important for a particular application, it may be important to reduce CPU cycles so as to consume less power; for applications running in small battery operated devices to huge data centres.
```
"from applications running..." instead of "for applications running..."

ablyx commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+## Using Appropriate Data Structures and Algorithms
+
+Using the appropriate Data Structures and Algorithms can improve the execution speed of your program trememdously. However, it requires time to learn and understand the nuances of each Data Structure and Algorithm and when to use them.
```
Using ~~the~~ appropriate...

ablyx commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+A good example of how Data Structures and Algorithms can improve execution time is with [binary search](https://www.tutorialspoint.com/data_structures_algorithms/binary_search_algorithm.htm). When finding a particular element in a list, you need to search through the entire list. If it has a million entries, every find will require you to look through **1,000,000** entries. However, if you have a ...
+
+If you are interested in learning more about Data Structures, you can learn more from the following resources:
```
"Data Structures and Algorithms" instead of just "Data Structures"?

ablyx commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+The faster a cache is, the more expensive it is. This means that the L1 cache (a few kilobytes big) is smaller than the L2 cache (a few megabytes big) and so on, until the RAM (a few gigabytes big).
+
+Whenever data is requested, the computer will first look in the cache for the data. If it exists, this is known as a **cache hit**. If it does not exist, this is known as a **cache miss**. When a **cache miss** happens, a contiguous block of memory (not just the requested data) is retrieved and stored onto the cache. Because of this, we want remember these rules to make cache friendly codes:
```
Because of this, we want to

ablyx commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+#### Row/Column Major Order
+
+This rule is how multidimensional arrays are stored in memory. Different programming languages have different methods of storing multidimensional arrays. [More Information](https://en.wikipedia.org/wiki/Row-_and_column-major_order).
```
This rule is about how

ablyx replied:
to be aware ~~what~~ of which major order

ablyx commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+#### Spatial Locality
+
+This rule states that memory stored near each other will likely be used in the near future. This means that using contiguous data structures, like arrays, helps improve the execution times. This is because the contiguous block of memory retrieved will likely contain the other memory that need to be used.
```
This means that using contiguous data structures, ~~like~~ such as arrays, helps improve execution times.

ablyx replied:
You might want to make the last sentence clearer. "contain the other memory that need to be used" sounds quite weird

ablyx commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+This rule states that memory stored near each other will likely be used in the near future. This means that using contiguous data structures, like arrays, helps improve the execution times. This is because the contiguous block of memory retrieved will likely contain the other memory that need to be used.
+
+An example of a Data Structure that does not do well in this aspect is [Linked List](https://www.tutorialspoint.com/data_structures_algorithms/linked_list_algorithms.htm). In a Linked List, each node can be stored anywhere on the memory. This means that there will likely be more cache misses when trying to iterate through a Linked List. This can cause Linked Lists to be much slower than what you  ...
```
~~This means that there will likely be more~~ Thus, you are likely to have more

ablyx commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+* Spatial Locality
+* Row/Column Major Order
+
+#### Temportal Locality
```
Temporal

ablyx reviewed (commented):
Good read! Maybe you can add a part about Algorithms for Using Appropriate Data Structures and Algorithms so that it is more complete.

IanTeo commented:
@ablyx thanks for the review. I have updated according to your review

ablyx reviewed (changes requested):
One more change - You forgot to change a typo, Temportal

Neurrone commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+Computer performance can be defined as the rate of work accomplished by a computer system. Even if execution time is not important for a particular application, it may be important to reduce CPU cycles so as to consume less power; from applications running in small battery operated devices to huge data centres.
+
+`Premature optimization is the root of all evil.`<br>
+I think many people have heard of this quote from Donald Knuth before. This quote is actually misintepreted frequently, because of the lack of context. Here is the full quote:<br>
```
misintepreted -> misinterpreted

Neurrone commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+To understand this topic, you only need to know how the cache works. It is okay if you do not understand exactly how the other components, registers, RAM and disk work.
+
+The faster a cache is, the more expensive it is. This means that the L1 cache (a few kilobytes big) is smaller than the L2 cache (a few megabytes big) and so on, until the RAM (a few gigabytes big).
```
Suggest adding a line at the start of this paragraph to explain what a cache is?

Neurrone commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,104 @@
+
+The faster a cache is, the more expensive it is. This means that the L1 cache (a few kilobytes big) is smaller than the L2 cache (a few megabytes big) and so on, until the RAM (a few gigabytes big).
+
+Whenever data is requested, the computer will first look in the cache for the data. If it exists, this is known as a **cache hit**. If it does not exist, this is known as a **cache miss**. When a **cache miss** happens, a contiguous block of memory (not just the requested data) is retrieved and stored onto the cache. Because of this, we want to remember these rules to make cache friendly codes:
```
Oh, cache explanation is here. Suggest swapping this paragraph and the previous - explaining what the cache is first would make the paragraph on different cache levels easier to understand.

IanTeo replied:
I don't think this is really an explanation of cache that is suitable to talk about before the previous line. The reason why I did not explain what a cache is properly is because I kind of expect the user to use the links I provided to learn what a cache is first.

Do you think I should just include an explanation anyway? Cause it is kind of hard to summarise an explanation of cache in a few lines

Neurrone replied:
Yeah, since cache friendly code is a major point to understand what you're saying, it should be there on the same page.

IanTeo commented:
@ablyx thanks for the review again. I have fixed the errors you pointed out
@Neurrone thank you for the suggestion. I have added a short explanation on cache. Do you think this is sufficient now?

ablyx reviewed (approved):
looks good to me!

Neurrone commented:
@IanTeo Yup, looks good.

damithc commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,102 @@
+
+Computer performance can be defined as the rate of work accomplished by a computer system. Even if execution time is not important for a particular application, it may be important to reduce CPU cycles so as to consume less power; from applications running in small battery operated devices to huge data centres.
+
+`Premature optimization is the root of all evil.`<br>
```
Use block quotes instead of cod formatting?

> Premature optimization is the root of all evil. --Donald Knuth

damithc commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,102 @@
+`We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%.`<br>
+You can find a nice explaination of the quote [here](http://wiki.c2.com/?PrematureOptimization).
+
+This is not a guide on how to optimize that 3%, but rather, to teach you standard optimization techniques that you should apply in all of your code, so that you do not create sub-optimal codes (premature pessimization). If you want to find out more about optimizing that 3%, you can find more at [Other Resources](#other-resources) below.
```
Use `describe` or `explain` instead of `teach`? The latter sets up unrealistic expectations.

damithc commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,102 @@
+* [Geeksforgeeks.org](http://www.geeksforgeeks.org/data-structures/)
+* [Introduction to Algorithms (Book)](https://www.amazon.com/Introduction-Algorithms-3rd-MIT-Press/dp/0262033844)
+* [MIT online course](http://courses.csail.mit.edu/6.851/spring12/lectures/)
+* [topcoder.com](https://www.topcoder.com/community/data-science/data-science-tutorials/)
```
Give a brief explanation of what the reader can expect to gain from each resource?

damithc replied:
Also, a more relevant resource is a one that talks about performance implications of data structures rather than links to general DSA resources.

damithc commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,102 @@
+* [MIT online course](http://courses.csail.mit.edu/6.851/spring12/lectures/)
+* [topcoder.com](https://www.topcoder.com/community/data-science/data-science-tutorials/)
+
+There are more of such resources online. Just Google and you shall find.
```
No need for this line.

damithc commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,102 @@
+
+To understand this topic, you only need to know how the cache works. It is okay if you do not understand exactly how the other components, registers, RAM and disk work. The cache is simply a place to store memory so that it can be accessed quickly. The cache is usually split into a few layers, L1, L2 and L3 cache, where the L1 cache is the fastest and L3 is the slowest. The faster a cache is, the ...
+
+Whenever data is requested, the computer will first look in the cache for the data. If it exists, this is known as a **cache hit**. If it does not exist, this is known as a **cache miss**. When a **cache miss** happens, a contiguous block of memory (not just the requested data) is retrieved and stored onto the cache. Because of this, we want to remember these rules to make cache friendly codes:
```
`a contiguous block of memory containing the requested data`?

damithc commented on contents/performance/Performance.md:
````diff
@@ -0,0 +1,102 @@
+System.out.println("Column major: " +(System.currentTimeMillis() - time));
+```
+
+In the example above, Row major takes around 100ms, while column major takes around 2000ms. You can use the codes above and try it on your own too.
````
>In the example above, Row major takes around 100ms, while column major takes around 2000ms.

These are the values you got by running the test in a typical notebook, right? Might want to mention that.

damithc commented on contents/performance/Performance.md:
```diff
@@ -0,0 +1,102 @@
+If you want to know more about Optimization, especially for that critical 3%, these other resources could be useful:
+
+* Finding the critical path - [Performance Profiling](PerformanceProfiling.md)
+* Speed up intensive computations - [GPGPU]()
```
Missing link

damithc reviewed (commented):
Good work  @IanTeo 
Some comments added. Revise as you see fit.
Adding more content and/or getting more peer reviews is welcome but not required. 
I will be merging end of week 12

damithc committed:
Merge branch 'master' into performance-introduction
